### DISCUSSION TITLE
Feature request: support for human-readable css class names

### DISCUSSION THREAD

**Discussion by resource11** [facebook/stylex#101] [General]
Hey y'all,

Thanks for all your work in creating this new CSS API! This tool will be better with the option to support human-readable CSS class names. I realize this is with the nebulous region of your architectural principles. However... [Users who rely on changing CSS classes for accessibility reasons right in the browser](https://ericwbailey.website/published/your-css-is-an-interface/) will be grateful you are considering them.

**Comment by necolas** [comment]
StyleX generates "atomic" or "utility" css rules, which means class names contain little more semantic information than inline styles would. Whether the class names are human readable or not, they can't reliably be used to target a specific UI component with style changes.

**Reply by derekjwilliams** [reply]
I know this thread is a year old, but there is a use case where having known classnames ahead of time would be beneficial, Lexical themes.  e.g. 

```
const lexicalTheme = {
  ltr: 'ltr',
  rtl: 'rtl',
  paragraph: 'editor-paragraph',
  ...
}
```

In the above, 'ltr', 'rtl', and 'editor-paragraph' are class names at run time.  For example: 

```
<p class="editor-paragraph"></p>
```

in the rendered html in a React application

I'd be curious how Meta handles this since they use (and created) both StlyeX and Lexical

**Reply by nmn** [reply]
You can use StyleX classes directly for setting Lexical themes:

```tsx
const lexicalTheme = {
  ltr: 'ltr',
  rtl: 'rtl',
  paragraph: stylex.props(styles.paragraph).className,
  ...
}
```

**NOTE:** You will have to be careful about not using Dynamic styles when doing this.

**Reply by necolas** [reply]
We style them with StyleX. Lexical has APIs that allow the relevant attributes to be set.

**Comment by nmn** [comment]
I know this wasn't the original question, but there's related questions here.

1. Perhaps what you actually want is the ability to apply additional class names that don't affect styling? In such scenarios, `aria` attributes are probably a better option. But if you 

2. You might want classNames for debugging in the browser. We provide the `dev` option to add classNames that tell you the `filename__variableName.keyname` for that. The class names that actually apply styles are hashes. We can probably consider using patterns for common styles in the future, but we would do that to shorten the strings and not for debuggability.

**Reply by resource11** [reply]
Thank you, @nmn! This is helpful. 

Option 2, outputting CSS class names like `filename__variableName.keyname` would work in a production environment. 

Here's why.

If you read the article I linked, you'd see how users would be able to target styles generated on pages created by styleX, and construct their own browser style sheets with particular overrides for accessibility needs.

Example: Site A created with styleX. Original page author isn't skilled at color contrast. Shipped to prod anyway.

User load up page. Notices styles generated with styleX and are human friendly classes.

This user targets the problematic styleX classes, and constructs their own custom browser stylesheets with particular overrides for **this one site** for their accessibility needs. 

Then, whenever visiting Site A, they can load up this custom stylesheet and be able to see the contrast. 

Hope that helps. Looking forward to this feature!

**Reply by nmn** [reply]
> outputting CSS class names like filename__variableName.keyname would work in a production environment.

I merged a change just before open sourcing that lets you enable the debugging classes without turning on dynamic style injection, so you can use it in production.

I think it makes sense to have a separate attribute with caconical IDs, but for lazy developers, this is the easiest thing to do.

**Reply by resource11** [reply]
Excellent. Is the enabling debugging classes feature documented somewhere? I'm currently responding in a mobile UI and don't have immediate access to check.

Re:
> I think it makes sense to have a separate attribute with caconical IDs, but for lazy developers, this is the easiest thing to do.



Can you define "lazy" here? You're breaking my neurospicy brain. 😅

**Reply by nmn** [reply]
> Is the enabling debugging classes feature documented somewhere?

Yeah, just set the `dev` option to `true` and the `runtimeInjection` option to `false`.

> Can you define "lazy"

This was a reference to your problem statement: developers shipping inaccessible UIs that end-users need to hack around. I was suggesting that it would make more sense to provide `data-id` attributes to make this easier for end-users, but that developers that don't care about a11y, probably wouldn't care about adding these `data-id` attributes either. The dev class names are automatic, so it's one fewer thing that you can forget to do.

**Reply by resource11** [reply]
Ahhh, I see.

Yes, using `data-id` attributes can be useful, depending on your CSS architectural strategy. YMMV. However. Even people who care about a11y and are **not** lazy by that definition can implement things with good intentions, and unintentionally make things harder to use or worse. That's a topic for another day.


Thanks for clarifying on the "lazy", and for pointing out where the debugging classes feature is documented. Appreciate your work.

**Reply by nmn** [reply]
I regret being spicy for no reason with the `lazy`.

**Reply by resource11** [reply]
Oh... you were being spicy. And, having a thread with... a neurospicy person (me). Who is Captain Literal when nuance is missing. Ooof. 😅 

Well, at least the outcome provided clarity and the necessary information to move forward!

**Comment by thevisioner** [comment]
Continuing discussion on Developer Experience, I'd like to provide some insights from my role as a front-end developer. My typical workflow involves creating layouts and styles, and subsequently testing them in the Chrome developer console. I recognize the significance of Atomic CSS, particularly in the context of large-scale projects, and I am genuinely eager to utilize a type-safe CSS solution that is compatible with React Server Components. However, I've found that my current workflow is not feasible (please refer to the attached screenshot for more details). And I also want to highlight.. in relation to CSS custom properties (variables). The meaning and context are not perceptible.

![image](https://github.com/facebook/stylex/assets/4502627/da59d608-64ec-40bc-8cf1-ec27c6e0e3e3)

I'm interested to know if you have any recommendations or if you've contemplated similar CSS development workflows in your planning. I suspect that this is not a situation unique to me, but rather a routine part of daily CSS work for many developers. I would greatly appreciate any thoughts or guidance on this matter.

**Reply by nmn** [reply]
Yes. This can be a bit of an adjustment. Here's my workflow when working *with* StyleX to build UIs:
1. Instead of editing existing CSS rule, work with inline styles (at the top of the sidebar)
2. Use the debug classNames to edit the source code directly.

There have been some prototypes for building dev tools to make this better, so I hope that'll exist one day.

**Reply by thevisioner** [reply]
Given the potential benefits to a developer's workflow, it may be worth exploring the feasibility of different CSS output configurations in development and production environments. Ideally, one would preserve the CSS style declaration block structure defined in the source code. Also, I think it would be useful to keep the CSS property (variable) names intact. Appreciate your insight.

**Reply by nmn** [reply]
> Ideally, one would preserve the CSS style declaration block structure defined in the source code.

This is not feasible. StyleX generates atomic styles (one style per class) and that is a core design detail that pretty much everything else depends on.

We are looking into a few ways to improve the debugging experience:
1. The debugging class names like `page__styles.desc` should show the relative file path instead like `app/about/page__styles.dec`
2. We're investigating if we can generate an *empty* CSS rule for these debug classNames which somehow contains a source map so that the declaration in JS can show up in dev tools.
3. Finally, there may eventually be a dev tools extension to give you a good and familiar experience.


> I think it would be useful to keep the CSS property (variable) names intact. 
Yes. This is valuable. Specially in development. I've already been thinking about this, and I think it's possible to implement.

---

NOTE: we do have higher priority tasks we need to tackle first.

**Reply by thevisioner** [reply]
Thanks for the extended answer. I appreciate your time and the work of all contributors. 💪
Still, I'm eager to see how the project develops. 🙏
